server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -929,6 +1015,11 @@ public static ClusterState readFrom(StreamInput in, DiscoveryNode localNode) thr
             // infer the versions from discoverynodes for now
             builder.nodes().getNodes().values().forEach(n -> builder.putTransportVersion(n.getId(), inferTransportVersion(n)));
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersion.current())) {
```

Comment:
I'd strongly recommend not using `TransportVersion.current()` for this, even in a draft PR - it'd be all too easy to miss one when finalizing things, and CI wouldn't be able to tell you about it until the next transport version bump (by which point the bad version may be released).